packages/relayer-lib/src/tx_builder/cosmos_to_cosmos.rs

```diff
@@ -0,0 +1,243 @@
+//! The `ChainSubmitter` submits txs to [`CosmosSdk`] based on events from [`CosmosSdk`].
+
+use anyhow::Result;
+use futures::future;
+use ibc_proto_eureka::{
+    ibc::{
+        core::{
+            channel::v2::{
+                Channel, MsgRecvPacket, MsgTimeout, QueryChannelRequest, QueryChannelResponse,
+            },
+            client::v1::{Height, MsgUpdateClient},
+        },
+        lightclients::tendermint::v1::ClientState,
+    },
+    Protobuf,
+};
+use prost::Message;
+use sp1_ics07_tendermint_utils::{light_block::LightBlockExt, rpc::TendermintRpcExt};
+use tendermint_rpc::{Client, HttpClient};
+
+use crate::{chain::CosmosSdk, events::EurekaEvent};
+
+use super::r#trait::TxBuilderService;
+
+/// The `TxBuilder` produces txs to [`CosmosSdk`] based on events from [`CosmosSdk`].
+#[allow(dead_code)]
+pub struct TxBuilder {
+    /// The HTTP client for the source chain.
+    pub source_tm_client: HttpClient,
+    /// The HTTP client for the target chain.
+    pub target_tm_client: HttpClient,
+    /// The signer address for the Cosmos messages.
+    pub signer_address: String,
+}
+
+impl TxBuilder {
+    /// Creates a new `TxBuilder`.
+    #[must_use]
+    pub const fn new(
+        source_tm_client: HttpClient,
+        target_tm_client: HttpClient,
+        signer_address: String,
+    ) -> Self {
+        Self {
+            source_tm_client,
+            target_tm_client,
+            signer_address,
+        }
+    }
+
+    /// Fetches the eureka channel state from the target chain.
+    /// # Errors
+    /// Returns an error if the channel state cannot be fetched or decoded.
+    pub async fn channel(&self, channel_id: String) -> Result<Channel> {
+        let abci_resp = self
+            .target_tm_client
+            .abci_query(
+                Some("/ibc.core.channel.v2.Query/Channel".to_string()),
+                QueryChannelRequest { channel_id }.encode_to_vec(),
+                None,
+                false,
+            )
+            .await?;
+
+        QueryChannelResponse::decode(abci_resp.value.as_slice())?
+            .channel
+            .ok_or_else(|| anyhow::anyhow!("No channel state found"))
+    }
+}
+
+#[async_trait::async_trait]
+impl TxBuilderService<CosmosSdk, CosmosSdk> for TxBuilder {
+    #[allow(clippy::too_many_lines)]
+    async fn relay_events(
```

Comment:
In general, I think this function should be split up more. I would say separate functions for each type of tx, or potentially target vs src events you are looking for + each event handling.

Reply:
I think this should also be done for eth_eureka, probably, but you could do that in a separate PR if you prefer.

Reply:
https://github.com/cosmos/solidity-ibc-eureka/issues/173

Reply:
Great, thanks!